File: src/com/mags/librarian/config/ConfigLoader.java
package com.mags.librarian.config;

import com.mags.librarian.Log;
import org.yaml.snakeyaml.Yaml;

import java.io.*;
import java.nio.file.FileAlreadyExistsException;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.logging.Level;
import java.util.stream.Collectors;

/**
 * Loads a configuration YAML file.
 */
public class ConfigLoader {

    private Yaml configYaml;
    private LinkedHashMap configObj;
    private Object configObj1;
    private ConfigAdaptor adaptor;

    public ConfigLoader() {
        configYaml = new Yaml();
    }

    /**
     * Return the raw configuration object.
     *
     * @return The raw configuration object.
     */
    public LinkedHashMap getConfigRaw() {
        return configObj;
    }

    /**
     * @param fileName
     * @throws FileNotFoundException
     */
    public void load(String fileName)
            throws FileNotFoundException {

        InputStream input = new FileInputStream(new File(fileName));

        this.configObj = (LinkedHashMap) this.configYaml.load(input);
    }

    /**
     * Load a configuration file contents from a string.
     *
     * @param document
     */
    public void loadFromString(String document) {
        this.configObj = (LinkedHashMap) this.configYaml.load(document);
    }

    public boolean createDefault(String templateFile, String fileName) throws IOException {

        File file = new File(fileName);
        if (file.exists()) {
            throw new FileAlreadyExistsException("File already exists. Delete or rename it an try again.");
        }

        // read default file

        InputStream res = getClass().getResourceAsStream(templateFile);
        BufferedReader br = new BufferedReader(new InputStreamReader(res));
        List<String> lines = br.lines().collect(Collectors.toList());

        // write it
        String content = String.join(System.lineSeparator(), lines);
        BufferedWriter bw = new BufferedWriter(new FileWriter(file.getAbsoluteFile()));
        bw.write(content);
        bw.close();

        return true;
    }

    /**
     * Retrieve a String value.
     *
     * @param key
     * @return The value of the key or null if not found.
     */
    protected String getValueString(String key) {
        return getValueString(key, null);
    }

    /**
     * Retrieve a String value with default.
     *
     * @param key
     * @param defaultValue
     * @return The value of the key or defaultValue if not found.
     */
    protected String getValueString(String key, String defaultValue) {

        String[] keyParts = key.split("\\.");

        LinkedHashMap map = this.configObj;
        String value = defaultValue;
        for (String keyPart : keyParts) {
            if (!map.containsKey(keyPart)) {
                return defaultValue;
            }

            value = map.get(keyPart).toString();
            if (map.get(keyPart).getClass().getName().equals("java.util.LinkedHashMap")) {
                map = (LinkedHashMap) map.get(keyPart);
            }
        }

        return value;
    }

    /**
     * Retrieve an Integer value.
     *
     * @param key
     * @return The value of the key or null if not found.
     */
    protected Integer getValueInt(String key) {
        return getValueInt(key, null);
    }

    /**
     * Retrieve an Integer value with default.
     *
     * @param key
     * @param defaultValue
     * @return The value of the key or defaultValue if not found.
     */
    protected Integer getValueInt(String key, Integer defaultValue) {

        String[] keyParts = key.split("\\.");

        LinkedHashMap map = this.configObj;
        Integer value = defaultValue;
        for (String keyPart : keyParts) {
            if (!map.containsKey(keyPart)) {
                return defaultValue;
            }

            value = Integer.parseInt(map.get(keyPart).toString());
            if (map.get(keyPart).getClass().getName().equals("java.util.LinkedHashMap")) {
                map = (LinkedHashMap) map.get(keyPart);
            }
        }

        return value;
    }

    /**
     * Retrieve a Boolean value.
     *
     * @param key
     * @return The value of the key or false if not found.
     */
    protected Boolean getValueBoolean(String key) {
        return getValueBoolean(key, false);
    }

    /**
     * * Retrieve a Boolean value with default.
     *
     * @param key
     * @param defaultValue
     * @return The value of the key or defaultValue if not found.
     */
    protected Boolean getValueBoolean(String key, boolean defaultValue) {

        String[] keyParts = key.split("\\.");

        LinkedHashMap map = this.configObj;
        boolean value = defaultValue;
        for (String keyPart : keyParts) {
            if (!map.containsKey(keyPart)) {
                return defaultValue;
            }

            value = Boolean.parseBoolean(map.get(keyPart).toString());
            if (map.get(keyPart).getClass().getName().equals("java.util.LinkedHashMap")) {
                map = (LinkedHashMap) map.get(keyPart);
            }
        }

        return value;
    }

    /**
     * Retrieve a list value.
     *
     * @param key
     * @return The value of the key as a List, or an empty List if not set.
     */
    protected List<String> getValueList(String key) {

        String[] keyParts = key.split("\\.");
        LinkedHashMap map = this.configObj;
        List<String> value = Collections.emptyList();
        for (String keyPart : keyParts) {
            if (!map.containsKey(keyPart)) {
                return Collections.emptyList();
            }

            if (map.get(keyPart) != null) {
                if (map.get(keyPart).getClass().getName().equals("java.util.ArrayList")) {
                    value = (List<String>) map.get(keyPart);
                }

                if (map.get(keyPart).getClass().getName().equals("java.util.LinkedHashMap")) {
                    map = (LinkedHashMap) map.get(keyPart);
                }
            }
        }

        return value;
    }

}

File: src/main/java/com/mags/librarian/config/ConfigLoader.kt
/*
 * This file is part of the librarian application.
 *
 * Copyright (c) Miguel Angel Gabriel <magabriel@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

package com.mags.librarian.config

import org.yaml.snakeyaml.Yaml
import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.util.*
import javax.inject.Inject

/**
 * Loads a configuration YAML file.
 */
class ConfigLoader
@Inject constructor(private val configYaml: Yaml) {

    private var onFileNotFoundCallback: ((fileName: String) -> Unit)? = null

    /**
     * Return the raw configuration object.
     */
    var configRaw: Map<String, Any>? = null
        private set
    private val configMap = LinkedHashMap<String, Any>()
    /**
     * Return the flattened configuration object.
     */
    internal val configFlat: Map<String, Any>
        get() = configMap

    fun load(fileName: String) {
        try {
            // load the config file
            val input = FileInputStream(File(fileName))
            @Suppress("UNCHECKED_CAST")
            configRaw = this.configYaml.load(input) as Map<String, Any>
            flatten("", configRaw, configMap)
        } catch (e: FileNotFoundException) {
            onFileNotFoundCallback?.invoke(fileName)
        }
    }

    /**
     * Load a configuration file contents from a string.
     */
    fun loadFromString(document: String) {
        @Suppress("UNCHECKED_CAST")
        configRaw = this.configYaml.load(document) as Map<String, Any>

        flatten("", configRaw, configMap)
    }

    /**
     * Flatten the config object (recursive hashmap) to a flat hasmap.
     */
    private fun flatten(baseKey: String,
                        recursiveMap: Map<String, Any>?,
                        flatMap: MutableMap<String, Any>) {
        recursiveMap!!.forEach { key, value ->
            var flatKey = key
            if (!baseKey.isEmpty()) {
                flatKey = baseKey + "." + key
            }

            if (value.javaClass.name == "java.util.LinkedHashMap") {
                @Suppress("UNCHECKED_CAST")
                flatten(flatKey, value as Map<String, Any>, flatMap)
            } else {
                flatMap.put(flatKey, value)
            }
        }
    }

    fun onFileNotFound(callback: (fileName: String) -> Unit) {
        onFileNotFoundCallback = callback
    }

    /**
     * Retrieve a String value with default.
     */
    @JvmOverloads internal fun getValueString(key: String,
                                              defaultValue: String = ""): String {
        @Suppress("UNCHECKED_CAST", "PLATFORM_CLASS_MAPPED_TO_KOTLIN")
        return (this.configMap as java.util.Map<String, Any>).getOrDefault(key,
                                                                           defaultValue) as String
    }

    /**
     * Retrieve an Integer value with default.
     */
    @JvmOverloads internal fun getValueInt(key: String,
                                           defaultValue: Int? = null): Int? {
        @Suppress("UNCHECKED_CAST", "PLATFORM_CLASS_MAPPED_TO_KOTLIN")
        return (this.configMap as java.util.Map<String, Any>).getOrDefault(key,
                                                                           defaultValue) as Int?
    }

    /**
     * * Retrieve a Boolean value with default.
     */
    @JvmOverloads internal fun getValueBoolean(key: String,
                                               defaultValue: Boolean? = false): Boolean? {
        @Suppress("UNCHECKED_CAST", "PLATFORM_CLASS_MAPPED_TO_KOTLIN")
        return (this.configMap as java.util.Map<String, Any>).getOrDefault(key,
                                                                           defaultValue) as Boolean
    }

    /**
     * Retrieve a list value (a list of Strings).
     */
    internal fun getValueListStrings(key: String): List<String> {
        return getValueListStrings(key, emptyList())
    }

    /**
     * Retrieve a list value (a list of Strings).
     */
    internal fun getValueListStrings(key: String,
                                     defaultValue: List<String>): List<String> {
        @Suppress("UNCHECKED_CAST", "PLATFORM_CLASS_MAPPED_TO_KOTLIN")
        return (this.configMap as java.util.Map<String, Any>).getOrDefault(key,
                                                                           defaultValue) as List<String>
    }

    /**
     * Retrieve a list value (a list of Maps)
     */
    internal fun getValueListMap(key: String): List<Map<String, Any>> {
        return getValueListMap(key, emptyList())
    }

    /**
     * Retrieve a list value (a list of Maps)
     */
    internal fun getValueListMap(key: String,
                                 defaultValue: List<Map<String, Any>>): List<Map<String, Any>> {
        @Suppress("UNCHECKED_CAST", "PLATFORM_CLASS_MAPPED_TO_KOTLIN")
        return (this.configMap as java.util.Map<String, Any>).getOrDefault(key,
                                                                           defaultValue) as List<Map<String, Any>>
    }
}

